WordPress/src/main/java/org/wordpress/android/ui/media/MediaGridFragment.java

```diff
@@ -306,6 +307,36 @@ private void setHasFetchedMediaForAllFilters() {
         }
     }
 
+    /*
+        given a list of media, it rechecks its state against the UploadingService to see
+        if it's uploading or queued and change it accordingly, to recover from an inconsistent state
+     */
+    private List<MediaModel> getSanitizedMedia(List<MediaModel> uncuredList) {
+        List<MediaModel> curedList = uncuredList;
+        Handler handler = new Handler();
+        for (final MediaModel media : uncuredList) {
+            if (MediaUploadState.QUEUED.toString().equals(media.getUploadState()) ||
+                    MediaUploadState.UPLOADING.toString().equals(media.getUploadState())) {
+                if (!UploadService.isPendingOrInProgressMediaUpload(media)) {
+                    // in order to avoid marking a genuine QUEUED item as FAILED as we hit this comparison
+                    // right in between the item being QUEUED and actually added to the queue, wait just a bit
+                    handler.postDelayed(new Runnable() {
+                        @Override
+                        public void run() {
+                            if (!UploadService.isPendingOrInProgressMediaUpload(media)) {
+                                // it is NOT being uploaded or queued in the actual UploadService, mark it failed
+                                media.setUploadState(MediaUploadState.FAILED);
```

Comment:
Since we're correcting the state in a delayed runnable, `curedList` will be returned by this method *before* the state is corrected. This ends up working because we dispatch `newUpdateMediaAction` here and that's intercepted by `onMediaChanged` in the activity which then takes care of updating the media item, but it seems fragile and inefficient. I think this is another sign it should be done elsewhere.